File: audiobook/src/main/java/de/ph1b/audiobook/adapter/FolderChooserAdapter.java
package de.ph1b.audiobook.adapter;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import java.io.File;
import java.util.List;

import butterknife.Bind;
import butterknife.ButterKnife;
import de.ph1b.audiobook.R;
import de.ph1b.audiobook.activity.FolderChooserActivity;


/**
 * Adapter for displaying files and folders.
 *
 * @author Paul Woitaschek
 */
public class FolderChooserAdapter extends BaseAdapter {

    private final Context c;
    private final List<File> data;
    private final FolderChooserActivity.OperationMode mode;

    /**
     * Constructor that initializes the class with the necessary values
     *
     * @param c    The context
     * @param data The files to show
     * @param mode The operation mode which defines the interaction.
     */
    public FolderChooserAdapter(@NonNull Context c, @NonNull List<File> data, FolderChooserActivity.OperationMode mode) {
        this.c = c;
        this.data = data;

        if (mode != FolderChooserActivity.OperationMode.SINGLE_BOOK &&
                mode != FolderChooserActivity.OperationMode.COLLECTION_BOOK) {
            throw new IllegalArgumentException("Invalid mode=" + mode);
        }
        this.mode = mode;
    }

    @Override
    public int getCount() {
        return data.size();
    }

    @Override
    @NonNull
    public File getItem(final int position) {
        return data.get(position);
    }

    @Override
    public long getItemId(final int position) {
        return position;
    }

    @Override
    public View getView(final int position, View convertView, final ViewGroup parent) {
        final ViewHolder viewHolder;
        if (convertView == null) {
            LayoutInflater li = LayoutInflater.from(c);
            convertView = li.inflate(R.layout.activity_folder_chooser_adapter_row_layout, parent,
                    false);

            viewHolder = new ViewHolder(convertView);
            convertView.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }

        File selectedFile = data.get(position);
        boolean isDirectory = selectedFile.isDirectory();

        viewHolder.textView.setText(selectedFile.getName());

        // if its not a collection its also fine to pick a file
        if (mode == FolderChooserActivity.OperationMode.COLLECTION_BOOK) {
            viewHolder.textView.setEnabled(isDirectory);
        }

        Drawable icon = ContextCompat.getDrawable(c, isDirectory ?
                R.drawable.ic_folder :
                R.drawable.ic_audiotrack_white_48dp);
        viewHolder.imageView.setImageDrawable(icon);
        viewHolder.imageView.setContentDescription(c.getString(isDirectory ?
                R.string.content_is_folder : R.string.content_is_file));

        return convertView;
    }

    static class ViewHolder {
        @Bind(R.id.singleline_text1) TextView textView;
        @Bind(R.id.singleline_image1) ImageView imageView;

        public ViewHolder(View parent) {
            ButterKnife.bind(this, parent);
        }
    }
}

File: audiobook/src/main/kotlin/de/ph1b/audiobook/adapter/FolderChooserAdapter.kt
package de.ph1b.audiobook.adapter

import android.content.Context
import android.support.v4.content.ContextCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.TextView
import de.ph1b.audiobook.R
import de.ph1b.audiobook.activity.FolderChooserActivity
import java.io.File


/**
 * Adapter for displaying files and folders.
 * Constructor that initializes the class with the necessary values
 *
 * @param c    The context
 * @param data The files to show
 * @param mode The operation mode which defines the interaction.
 *
 * @author Paul Woitaschek
 */
class FolderChooserAdapter(private val c: Context, private val data: List<File>, private val mode: FolderChooserActivity.OperationMode) : BaseAdapter() {

    init {
        if (mode !== FolderChooserActivity.OperationMode.SINGLE_BOOK && mode !== FolderChooserActivity.OperationMode.COLLECTION_BOOK) {
            throw IllegalArgumentException("Invalid mode=" + mode)
        }
    }

    override fun getCount(): Int {
        return data.size
    }

    override fun getItem(position: Int): File {
        return data[position]
    }

    override fun getItemId(position: Int): Long {
        return position.toLong()
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var view = convertView
        val viewHolder: ViewHolder
        if (view == null) {
            val li = LayoutInflater.from(c)
            view = li.inflate(R.layout.activity_folder_chooser_adapter_row_layout, parent,
                    false)

            viewHolder = ViewHolder(view)
            view.tag = viewHolder
        } else {
            viewHolder = view.tag as ViewHolder
        }

        val selectedFile = data[position]
        val isDirectory = selectedFile.isDirectory

        viewHolder.textView.text = selectedFile.name

        // if its not a collection its also fine to pick a file
        if (mode === FolderChooserActivity.OperationMode.COLLECTION_BOOK) {
            viewHolder.textView.isEnabled = isDirectory
        }

        val icon = ContextCompat.getDrawable(c, if (isDirectory) R.drawable.ic_folder else R.drawable.ic_audiotrack_white_48dp)
        viewHolder.imageView.setImageDrawable(icon)
        viewHolder.imageView.contentDescription = c.getString(if (isDirectory) R.string.content_is_folder else R.string.content_is_file)

        return view!!
    }

    internal class ViewHolder(parent: View) {
        internal val textView: TextView
        internal val imageView: ImageView

        init {
            textView = parent.findViewById(R.id.singleline_text1) as TextView
            imageView = parent.findViewById(R.id.singleline_image1) as ImageView
        }
    }
}
